pkg/boot/grub/grub.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2017-2020 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package grub implements a grub config file parser.
+//
+// See the grub manual https://www.gnu.org/software/grub/manual/grub/ for
+// a reference of the configuration format
+// In particular the following pages:
+// - https://www.gnu.org/software/grub/manual/grub/html_node/Shell_002dlike-scripting.html
+// - https://www.gnu.org/software/grub/manual/grub/html_node/Commands.html
+//
+// Currently, only the linux[16|efi], initrd[16|efi], menuentry and set
+// directives are partially supported.
+package grub
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/url"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/boot"
+	"github.com/u-root/u-root/pkg/boot/multiboot"
+	"github.com/u-root/u-root/pkg/curl"
+	"github.com/u-root/u-root/pkg/uio"
+)
+
+var (
+	// ErrDefaultEntryNotFound is returned when the configuration file
+	// names a default label that is not part of the configuration.
+	ErrDefaultEntryNotFound = errors.New("default variable not set in configuration")
+	// ErrInitrdUsedWithoutLinux is returned when an initrd directive is
+	// not following a linux directive in the same menu entry
+	ErrInitrdUsedWithoutLinux = errors.New("missing linux directive before initrd")
+	// ErrModuleUsedWithoutMultiboot is returned when a module directive is
+	// not following a multiboot directive in the same menu entry
+	ErrModuleUsedWithoutMultiboot = errors.New("missing multiboot directive before module")
+)
+
+// Config encapsulates a parsed grub configuration file.
+type Config struct {
```

Comment:
I noticed that pkg/boot/syslinux/syslinux.go seems to redefine a couple of these common elements like `type Config` and `type parser` - would it make sense to combine these into a file boot_parsers.go or something of common parser definitions to dedup?

Reply:
for the moment, I think it's okay not to -- I have some changes on the way to this part of the infrastructure